### PR TITLE
Painless: Move More Logic to PainlessLookup

### PR DESCRIPTION
**Pull request by jdconrad**
This moves some run-time lookups for methods and fields to the PainlessLookup.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java [hunk 1]
```diff
@@ -460,46 +395,28 @@ static MethodHandle lookupGetter(PainlessLookup painlessLookup, Class<?> receive
...
-        for (Class<?> clazz = receiverClass; clazz != null; clazz = clazz.getSuperclass()) {
-            PainlessClass struct = painlessLookup.lookupPainlessClass(clazz);
-
-            if (struct != null) {
-                MethodHandle handle = struct.setterMethodHandles.get(name);
-                if (handle != null) {
-                    return handle;
+        try {
+            return painlessLookup.lookupRuntimeSetterMethodHandle(receiverClass, name);
+        } catch (IllegalArgumentException iae) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: I don't think we should use IAE as control flow. Can we return null if the the class is not found and base these special cases on that?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookup.java [hunk 2]
```diff
@@ -174,4 +177,76 @@ public PainlessMethod lookupFunctionalInterfacePainlessMethod(Class<?> targetCla
 
         return functionalInterfacePainlessMethod;
     }
+
+    public PainlessMethod lookupRuntimePainlessMethod(Class<?> originalTargetClass, String methodName, int methodArity) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookup.java]: Can you add at least a single line javadoc to these methods (and preferably the other ones as well here, although that could be in a followup)? It is getting difficult to understand what the differences in expected behavior are, eg based on a single word difference here of adding "runtime" to the method name, but taking the same args as `lookupPainlessMethod` above.

**Review by rjernst — COMMENTED** [review]
Looks good, but I think we should return null instead of using exceptions to indicate a class is not found.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: Fixed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookup.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookup.java]: I'd like to add documentation for all of these in a follow up PR as there are a few more changes lined up and I would prefer not to write it twice over the next couple weeks if that's okay.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review.  I changed all the methods in PainlessLookup to return null and react at the site of the calls instead of using exception handling as control flow.  Ready for another round.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by jdconrad** [comment]
@rjernst Thanks for the reviews!

**Pull request merged**